Trust user-added CAs on Android

shrizza opened the pull request:
In response to issue #3312. In Android, allow trusting user-added CAs, while still deferring trust to public CAs for both zulipchat.com and zulip.chat.org.

Fixes: #3312

chrisbobbe commented:
Hi @shrizza, thanks for the PR! `gnprice` will need to approve it before it's merged, and he's on vacation until next Tuesday Pacific time, 2022-09-22. 🙂

These lines look like they're meant to identify the domains Greg pointed to in his [proposal](https://github.com/zulip/zulip-mobile/issues/3312#issuecomment-525952536):

```xml
<domain includeSubdomains="true">zulipchat.com</domain>
```
```xml
<domain includeSubdomains="true">zulip.org</domain>
```

Those domains were: "zulipchat.com and its subdomains, and chat.zulip.org." For the latter, it looks like your code actually points to zulip.org and all its subdomains. Is there a reason you chose to do that instead of just "chat.zulip.org"? Thanks!

Also, it would help if you [squashed](https://zulip.readthedocs.io/en/latest/git/using.html#examine-and-tidy-your-commit-history) your branch's two commits into a single commit and gave that a commit message that follows our [style guide](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages). But for one small PR, we also don't mind doing that ourselves.

Oh, and a question for Greg when you get back: should we also think about zulip.com and its subdomains?

shrizza commented:
Hi @chrisbobbe,

Thank you for the feedback. I've squashed the changes and modified the commit message. Should this PR's title be adjusted similarly?

I'm actually not familiar with how the zulipchat.com and zulip.org domains are organized, but if there's an elaborate subdomain delegation scheme I can adjust the deferred domains to strictly zulipchat.com and chat.zulip.org, and no further subdomains. I will wait on Greg's response on this point so that it can all be considered together with the verdict on zulip.com.

chrisbobbe commented:
(Tagging @gnprice since I didn't do so above. 🙂)

gnprice commented on android/app/src/main/res/xml/network_security_config.xml:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="utf-8"?>
+<network-security-config>
+      <base-config>
+           <trust-anchors>
+                <certificates src="system" />
+                <certificates src="user" />
+           </trust-anchors>
```
I'd like to have a comment here with a sentence or two explaining what this is doing and why, and also one for the `domain-config` element. Should link to the request https://github.com/zulip/zulip-mobile/issues/3312 , and also to the reference docs for understanding this file's semantics:
https://developer.android.com/training/articles/security-config#FileFormat

I'll probably write those comments up in the course of merging this. But if you can make a go at writing a draft, that would help me out and would be appreciated.

shrizza replied:
I've added a number of comments to address your concern. Please let me know if it needs any further adjustment.

gnprice reviewed (commented):
Thanks @shrizza for sending this! Sorry for the slow response -- I came down with covid the day after getting back from vacation, and then when I emerged, this was a bit buried in the pile of other things to catch up on.

Generally this looks good -- small comments below.

One question for you: have you been able to try this out in a build of the app, and confirmed that it succeeds in solving your use case? Based on the documentation I believe it should, but it'd be great to have empirical confirmation that we haven't missed something.

gnprice commented:
Oh, and one more small comment: please mark the commit and the PR as resolving the issue. See:
https://github.com/zulip/zulip-mobile/blob/main/docs/style.md#mention-fixed-issue
https://github.com/zulip/zulip-mobile/blob/main/docs/style.md#fixes-format

shrizza commented on android/app/src/main/res/xml/network_security_config.xml:
```diff
@@ -0,0 +1,17 @@
+           <trust-anchors><certificates src="system" /></trust-anchors>
+      </domain-config>
+      <domain-config>
+           <domain includeSubdomains="true">zulip.org</domain>
```
Yes, applying this policy uniformly across all those domains should be fine as long as you do not find yourself delegating subdomains of those to third-parties that may wish to use private CAs.

shrizza replied:
PR is updated to additionally include zulip.com.

shrizza commented on android/app/src/main/res/xml/network_security_config.xml:
```diff
@@ -0,0 +1,17 @@
+      </domain-config>
+      <domain-config>
+           <domain includeSubdomains="true">zulip.org</domain>
+           <trust-anchors><certificates src="system" /></trust-anchors>
```
I've adjusted the PR to fulfill this desire.

shrizza commented:
> Oh, and one more small comment: please mark the commit and the PR as resolving the issue. See: https://github.com/zulip/zulip-mobile/blob/main/docs/style.md#mention-fixed-issue https://github.com/zulip/zulip-mobile/blob/main/docs/style.md#fixes-format

I've adjusted the commit message accordingly.

I'm not sure about the PR since it seems I've referenced the issue in both the title and my first comment. I'm not seeing a description I can edit for this PR; please let me know if I'm missing something.

shrizza commented:
> One question for you: have you been able to try this out in a build of the app, and confirmed that it succeeds in solving your use case? Based on the documentation I believe it should, but it'd be great to have empirical confirmation that we haven't missed something.

To be honest, I have no experience in Android app development and have not actually tested my changes. I don't have an Android SDK set up either so I'd imagine that would be necessary for testing.

gnprice commented:
Thanks for the revision!

> To be honest, I have no experience in Android app development and have not actually tested my changes. I don't have an Android SDK set up either so I'd imagine that would be necessary for testing.

Perfectly reasonable. Yeah, the app should be pretty straightforward to build once one has an Android dev setup, but the latter can be a bit annoying and a few tens of GB of disk.

Before merge I'll test locally that it builds, but it would take a fair bit of setup to directly test that it solves the issue. We'll just rely, then, on the fact it's a short file and the docs seem pretty clear about its semantics.



> I'm not sure about the PR since it seems I've referenced the issue in both the title and my first comment. I'm not seeing a description I can edit for this PR; please let me know if I'm missing something.

Ah, I see how the style guide isn't clear about this, thanks. I'll go try to clarify there… OK, please take a look now:
https://github.com/zulip/zulip-mobile/blob/main/docs/style.md#mention-fixed-issue
or to see just what changed: https://github.com/zulip/zulip-mobile/commit/4e09f23b9dbaca238f0c76fa96a77b46e5d5f2ba

Do let me know if that still doesn't do it. We aspire to have the docs clear enough to answer this sort of recurring question by themselves, so this feedback when they aren't is helpful.

gnprice commented:
> Before merge I'll test locally that it builds

(We test this in CI, but CI failed here -- [following the link](https://github.com/zulip/zulip-mobile/actions/runs/3356241599/jobs/5561161514), it looks like an operational issue unrelated to your change:
```
* What went wrong:
A problem occurred configuring project ':react-native-photo-view'.
> Could not resolve all artifacts for configuration ':react-native-photo-view:classpath'.
   > Could not resolve com.android.tools.build:gradle-core:2.1.0.
     Required by:
         project :react-native-photo-view > com.android.tools.build:gradle:2.1.0
      > Could not resolve com.android.tools.build:gradle-core:2.1.0.
         > Could not get resource 'https://jcenter.bintray.com/com/android/tools/build/gradle-core/2.1.0/gradle-core-2.1.0.pom'.
            > Could not GET 'https://jcenter.bintray.com/com/android/tools/build/gradle-core/2.1.0/gradle-core-2.1.0.pom'.
               > Read timed out
```
but which caused it to fail before the actual Android build. I guess I'll also rerun CI to see if that makes it pass. Previous instance at https://github.com/zulip/zulip-mobile/pull/5193#issuecomment-1011472231; [chat thread](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/jcenter.20shutdown/near/1309290) from then.)

shrizza commented:
> Do let me know if that still doesn't do it. We aspire to have the docs clear enough to answer this sort of recurring question by themselves, so this feedback when they aren't is helpful.
Thank you for expanding on the meaning of the PR description; that does make it more clear to me. I've adjusted this PR's title and description to more closely fit the guideline.

gnprice commented:
Hmm, well, I'm glad I tested this a bit manually before merging it -- because in its current form, it causes the app to stop working in the debug builds we use for development!

I suspect what's happening is that it's inadvertently clobbering the debug-only config that allows plaintext HTTP traffic, which the debug builds use for talking to the local web server that provides the app's JS code.

I'll do a bit more poking at it now to see if I can resolve the issue. In the meantime, I rebased and edited the comments to a version I'm happy with.

gnprice commented:
> I suspect what's happening is that it's inadvertently clobbering the debug-only config that allows plaintext HTTP traffic, which the debug builds use for talking to the local web server that provides the app's JS code.

Yup, that's it:
https://developer.android.com/guide/topics/manifest/application-element#usesCleartextTraffic
> This flag is ignored on Android 7.0 (API level 24) and above if an Android Network Security Config is present.

I think the fix will be straightforward.

gnprice commented:
OK, fixed that -- see https://github.com/zulip/zulip-mobile/commit/85c3a716a7bbf00c936beaf34bc84a0e6bf5c885 .

And merged! Thanks again @shrizza for the contribution.
